cli/src/commands/graph/federated-graph/utils.ts

```diff
@@ -0,0 +1,240 @@
+  client,
+  name,
+  namespace,
+}: {
+  client: Client;
+  name: string;
+  namespace: string;
+}) => {
+  const resp = await client.platform.getFederatedGraphByName(
+    {
+      name,
+      namespace,
+      includeMetrics: false,
+    },
+    {
+      headers: getBaseHeaders(),
+    },
+  );
+
+  if (resp.response?.code !== EnumStatusCode.OK) {
+    console.log(`${pc.red(`Could not fetch the federated graph ${pc.bold(name)}`)}`);
+    if (resp.response?.details) {
+      console.log(pc.red(pc.bold(resp.response?.details)));
+    }
+    process.exit(1);
+  }
+
+  const subgraphs = await resp.subgraphs;
+
+  return subgraphs.map((s) => {
+    return {
+      name: s.name,
+      routingURL: s.routingURL,
+      subscriptionURL: s.subscriptionUrl,
+      subscriptionProtocol: s.subscriptionProtocol,
+      isV2Graph: s.isV2Graph,
+    };
+  });
+};
+
+export const getFederatedGraphSDL = async ({
+  client,
+  name,
+  namespace,
+}: {
+  client: Client;
+  name: string;
+  namespace: string;
+}) => {
+  const resp = await client.platform.getFederatedGraphSDLByName(
+    {
+      name,
+      namespace,
+    },
+    {
+      headers: getBaseHeaders(),
+    },
+  );
+
+  if (resp.response?.code !== EnumStatusCode.OK || !resp?.sdl) {
+    console.log(`${pc.red(`Could not fetch the SDL of the federated graph ${pc.bold(name)}`)}`);
+    if (resp.response?.details) {
+      console.log(pc.red(pc.bold(resp.response?.details)));
+    }
+    process.exit(1);
+  }
+
+  const sdl = await resp.sdl;
+
+  return sdl;
+};
+
+// Returns the latest valid schema version of a subgraph that was composed to form the provided federated graph.
+export const getSubgraphSDL = async ({
+  client,
+  fedGraphName,
+  subgraphName,
+  namespace,
+}: {
+  client: Client;
+  fedGraphName: string;
+  subgraphName: string;
+  namespace: string;
+}) => {
+  const resp = await client.platform.getSubgraphSDLFromLatestComposition(
+    {
+      name: subgraphName,
+      namespace,
+      fedGraphName,
+    },
+    {
+      headers: getBaseHeaders(),
+    },
+  );
+
+  if (resp.response?.code !== EnumStatusCode.OK) {
+    console.log(`${pc.red(`Could not fetch the SDL of the subgraph ${pc.bold(subgraphName)}`)}`);
+    if (resp.response?.details) {
+      console.log(pc.red(pc.bold(resp.response?.details)));
+    }
+    process.exit(1);
+  }
+
+  const sdl = await resp.sdl;
+
+  return sdl;
+};
+
+export const injectRequiredDirectives = (schema: string, isV2Graph?: boolean) => {
+  const linkDirective = `extend schema
+  @link(
+    url: "https://specs.apollo.dev/federation/v2.5"
+    import: [
+      "@authenticated"
+      "@composeDirective"
+      "@extends"
+      "@external"
+      "@inaccessible"
+      "@interfaceObject"
+      "@override"
+      "@provides"
+      "@key"
+      "@requires"
+      "@requiresScopes"
+      "@shareable"
+      "@tag"
+    ]
+  )
+
+`;
+  let hasDefinedLink = false;
+  if (!isV2Graph) {
+    return schema;
+  }
+  visit(parse(schema), {
```

Comment:
parse can fail, but I guess at this point we must have a valid schema

Reply:
yep